packages/snaps-rpc-methods/src/permitted/getAllSnaps.ts

```diff
@@ -0,0 +1,71 @@
+import type { JsonRpcEngineEndCallback } from '@metamask/json-rpc-engine';
+import type { PermittedHandlerExport } from '@metamask/permission-controller';
+import { rpcErrors } from '@metamask/rpc-errors';
+import type { GetSnapsResult } from '@metamask/snaps-sdk';
+import type {
+  JsonRpcParams,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+
+import type { MethodHooksObject } from '../utils';
+
+const hookNames: MethodHooksObject<GetAllSnapsHooks> = {
+  getAllSnaps: true,
+};
+
+/**
+ * `wallet_getAllSnaps` gets all installed Snaps. Currently, this can only be
+ * called from `https://snaps.metamask.io`.
+ */
+export const getAllSnapsHandler: PermittedHandlerExport<
+  GetAllSnapsHooks,
+  JsonRpcParams,
+  GetSnapsResult
+> = {
+  methodNames: ['wallet_getAllSnaps'],
+  implementation: getAllSnapsImplementation,
+  hookNames,
+};
+
+export type GetAllSnapsHooks = {
+  /**
+   * @returns All installed Snaps.
+   */
+  getAllSnaps: () => Promise<GetSnapsResult>;
+};
+
+/**
+ * The `wallet_getSnaps` method implementation.
+ * Fetches available snaps for the requesting origin and adds them to the JSON-RPC response.
+ *
+ * @param request - The JSON-RPC request object.
+ * @param response - The JSON-RPC response object.
+ * @param _next - The `json-rpc-engine` "next" callback. Not used by this
+ * function.
+ * @param end - The `json-rpc-engine` "end" callback.
+ * @param hooks - The RPC method hooks.
+ * @param hooks.getAllSnaps - A function that returns all installed snaps.
+ * @returns Nothing.
+ */
+async function getAllSnapsImplementation(
+  request: JsonRpcRequest,
+  response: PendingJsonRpcResponse<GetSnapsResult>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { getAllSnaps }: GetAllSnapsHooks,
+): Promise<void> {
+  // The origin is added by the MetaMask middleware stack.
+  const { origin, method } = request as JsonRpcRequest & { origin: string };
+
+  if (origin !== 'https://snaps.metamask.io') {
+    return end(
+      rpcErrors.invalidRequest({
```

Comment:
Should we throw method not found instead?

Reply:
We throw this for `wallet_invokeKeyring`, but I can change it if you prefer.

Reply:
I think I would prefer "hiding" it from everyone else other than the correct origin for now.

Similarly to what we do for the `snap_` methods: https://github.com/MetaMask/snaps/blob/main/packages/snaps-rpc-methods/src/permitted/middleware.ts#L27-L31